test/hotspot/jtreg/compiler/lib/generators/SpecialDoubleGenerator.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.lib.generators;
+
+import java.util.Random;
+import jdk.test.lib.Utils;
+
+/**
+ * Provide a double distribution picked from a list of special values, including NaN, zero, int, etc.
+ */
+public final class SpecialDoubleGenerator extends DoubleGenerator {
```

Comment:
Same as for MixedLong/IntGenerator: Special*Generator could share common logic in a superclass if Double/FloatGenerator are interfaces.

Reply:
What I'm thinking of doesn't involve any generics. I'm talking about specialMinFrequency, specialMaxFrequency, specialCountDown and the logic in next* to decide whether to provide a special value or not now. This is independent from whether float or double is used, as far as I can tell.

Reply:
Ok. I suppose I could factor this out into some extra class and extra file.
But is that worth it? Then I need to still somehow route all that logic through the constructor, and `nextDouble`. I think in the end I have more code and more abstractions, I don't yet see how that is worth it.

Maybe you had a better idea in mind that I did not yet think of ;)

Reply:
To be more clear, without generics:

```java
abstract class FloatingGenerator {
      // specialCountDown detemines in how many iterations we generate the next special value.
     private final int specialMinFrequency;
     private final int specialMaxFrequency;
     private int specialCountDown;

    // .. constructor ...

     protected final bool shouldUseSpecial() {
         specialCountDown--;
         if (specialCountDown <= 0) {
             specialCountDown = RANDOM.nextInt(specialMinFrequency, specialMaxFrequency);
             return true;
         } else {
             return false;
         }  
     }
}
```

Then in SpecialDoubleGenerator:

```java
     @Override
     public double nextDouble() {
         if (shouldUseSpecial()) {
             int r = RANDOM.nextInt(VALUES.length);
             return VALUES[r];
         } else {
             return backgroundGenerator.nextDouble();
         }
     }
```

If you make FloatingGenerator generic, though, of course, and rename the methods from nextDouble and nextFloat to just next for the generators, then everything except for the list of special values itself could be moved into FloatingGenerator. That would result in very clean code in my opinion. If you need to adjust the logic for special selection later, with the current approach, that requires changes in multiple locations, which in my experience, ultimately leads to inconsistencies.

Reply:
I posted my second comment without seeing your latest comment. So it's not really a reply to that.

Reply:
Hmm. I would have to make the classes `FloatGenerator` and `DoubleGenerator` interfaces. But having default methods for interfaces is not really the recommended way, as far as I know. And then I cannot make those default methods final, which sucks - java does not allow it.

Alternative: I make a separate class `RandomDistance` to compose into the others, that one can query with `isDistanceReached`. Would that be worth it you think?